{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "86d54fee",
   "metadata": {},
   "source": [
    "### Small script which helps converting sample distribution to avg genome coverage\n",
    "\n",
    "Perhaps we should also include the actual cluster coverage from the BAM files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "47e6ced3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>distribution</th>\n",
       "      <th>samle</th>\n",
       "      <th>ncbi</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>weighted_dist</th>\n",
       "      <th>readshare</th>\n",
       "      <th>expected_average_coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NC_014328_1</td>\n",
       "      <td>4630065</td>\n",
       "      <td>926013.0</td>\n",
       "      <td>8.238640e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_CP020566_1</td>\n",
       "      <td>2071952</td>\n",
       "      <td>414390.4</td>\n",
       "      <td>3.686788e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome3</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_CP053893_1</td>\n",
       "      <td>5984367</td>\n",
       "      <td>1196873.4</td>\n",
       "      <td>1.064846e+08</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_LT906445_1</td>\n",
       "      <td>2132142</td>\n",
       "      <td>426428.4</td>\n",
       "      <td>3.793888e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_LT906470_1</td>\n",
       "      <td>2041290</td>\n",
       "      <td>408258.0</td>\n",
       "      <td>3.632228e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NC_014328_1</td>\n",
       "      <td>4630065</td>\n",
       "      <td>926013.0</td>\n",
       "      <td>8.238640e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_CP020566_1</td>\n",
       "      <td>2071952</td>\n",
       "      <td>414390.4</td>\n",
       "      <td>3.686788e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome3</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_CP053893_1</td>\n",
       "      <td>5984367</td>\n",
       "      <td>1196873.4</td>\n",
       "      <td>1.064846e+08</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_LT906445_1</td>\n",
       "      <td>2132142</td>\n",
       "      <td>426428.4</td>\n",
       "      <td>3.793888e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_LT906470_1</td>\n",
       "      <td>2041290</td>\n",
       "      <td>408258.0</td>\n",
       "      <td>3.632228e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NC_014328_1</td>\n",
       "      <td>4630065</td>\n",
       "      <td>926013.0</td>\n",
       "      <td>8.238640e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_CP020566_1</td>\n",
       "      <td>2071952</td>\n",
       "      <td>414390.4</td>\n",
       "      <td>3.686788e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome3</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_CP053893_1</td>\n",
       "      <td>5984367</td>\n",
       "      <td>1196873.4</td>\n",
       "      <td>1.064846e+08</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_LT906445_1</td>\n",
       "      <td>2132142</td>\n",
       "      <td>426428.4</td>\n",
       "      <td>3.793888e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_LT906470_1</td>\n",
       "      <td>2041290</td>\n",
       "      <td>408258.0</td>\n",
       "      <td>3.632228e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id  distribution     samle           ncbi  genome_size  weighted_dist  \\\n",
       "0  Genome1           0.2  sample_0    NC_014328_1      4630065       926013.0   \n",
       "1  Genome2           0.2  sample_0  NZ_CP020566_1      2071952       414390.4   \n",
       "2  Genome3           0.2  sample_0  NZ_CP053893_1      5984367      1196873.4   \n",
       "3  Genome4           0.2  sample_0  NZ_LT906445_1      2132142       426428.4   \n",
       "4  Genome5           0.2  sample_0  NZ_LT906470_1      2041290       408258.0   \n",
       "0  Genome1           0.2  sample_2    NC_014328_1      4630065       926013.0   \n",
       "1  Genome2           0.2  sample_2  NZ_CP020566_1      2071952       414390.4   \n",
       "2  Genome3           0.2  sample_2  NZ_CP053893_1      5984367      1196873.4   \n",
       "3  Genome4           0.2  sample_2  NZ_LT906445_1      2132142       426428.4   \n",
       "4  Genome5           0.2  sample_2  NZ_LT906470_1      2041290       408258.0   \n",
       "0  Genome1           0.2  sample_1    NC_014328_1      4630065       926013.0   \n",
       "1  Genome2           0.2  sample_1  NZ_CP020566_1      2071952       414390.4   \n",
       "2  Genome3           0.2  sample_1  NZ_CP053893_1      5984367      1196873.4   \n",
       "3  Genome4           0.2  sample_1  NZ_LT906445_1      2132142       426428.4   \n",
       "4  Genome5           0.2  sample_1  NZ_LT906470_1      2041290       408258.0   \n",
       "\n",
       "      readshare  expected_average_coverage  \n",
       "0  8.238640e+07                  17.793788  \n",
       "1  3.686788e+07                  17.793788  \n",
       "2  1.064846e+08                  17.793788  \n",
       "3  3.793888e+07                  17.793788  \n",
       "4  3.632228e+07                  17.793788  \n",
       "0  8.238640e+07                  17.793788  \n",
       "1  3.686788e+07                  17.793788  \n",
       "2  1.064846e+08                  17.793788  \n",
       "3  3.793888e+07                  17.793788  \n",
       "4  3.632228e+07                  17.793788  \n",
       "0  8.238640e+07                  17.793788  \n",
       "1  3.686788e+07                  17.793788  \n",
       "2  1.064846e+08                  17.793788  \n",
       "3  3.793888e+07                  17.793788  \n",
       "4  3.632228e+07                  17.793788  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "\n",
    "def generate_simulation_overview(dir_dataset:str):\n",
    "    #Get genome sizes\n",
    "    fp_combined_input_genomes = \"../data/simulated_data/input_genomes/combined.fa\" #TODO: fix hardcoded?\n",
    "    genome_size = {}\n",
    "    for record in SeqIO.parse(fp_combined_input_genomes, \"fasta\"):\n",
    "        genome_size[record.name.replace(\".\",\"_\")] = len(record.seq)\n",
    "        \n",
    "    #Get mapping between camisim id and ncbi ids\n",
    "    fp_map = os.path.join(dir_dataset, \"internal/genome_locations.tsv\")\n",
    "    cam2genom = {}\n",
    "    for line in open(fp_map, \"r\"):\n",
    "        genomeid, fp = line.strip().split(\"\\t\")\n",
    "        ncbi_id = os.path.basename(fp).rsplit(\".\",1)[0]\n",
    "        cam2genom[genomeid] = ncbi_id\n",
    "        cam2genom[ncbi_id] = genomeid\n",
    "        \n",
    "    #Run over the samples and calculate the expected coverage from genome size and distribution.\n",
    "    dfs = []\n",
    "    for fp_dist in glob.glob(os.path.join(dir_dataset, \"distributions/*\")):\n",
    "        df = pd.read_csv(fp_dist, sep=\"\\t\", names=[\"id\", \"distribution\"])\n",
    "        df[\"samle\"] = \"sample_\" + fp_dist.rsplit(\"_\",1)[1].replace(\".txt\",\"\")\n",
    "        df[\"ncbi\"] = [cam2genom[x] for x in df.id]\n",
    "        df[\"genome_size\"] = [genome_size[x] for x in df.ncbi]\n",
    "        df[\"weighted_dist\"] = df.distribution * df.genome_size\n",
    "        df[\"readshare\"] = (df.weighted_dist/df.weighted_dist.sum())*(0.3*10**9)\n",
    "        df[\"expected_average_coverage\"] = df.readshare / df.genome_size\n",
    "\n",
    "        dfs.append(df)\n",
    "    df_total = pd.concat(dfs)\n",
    "    return df_total\n",
    "\n",
    "generate_simulation_overview(\"../data/simulated_data/camisim/0_4GB/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d44c3191",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.4"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "float(os.path.basename(\"../data/simulated_data/camisim/0_4GB/\".rstrip(\"/\").replace(\"GB\",\"\").replace('_','.')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "b17dbf63",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_average_coverages(dir_datalabel:str):\n",
    "    \"\"\"\n",
    "    cov_i = (pi/(sum(pj,sj)))*totalBp\n",
    "    \"\"\"\n",
    "    fp = os.path.join(dir_datalabel, \"internal/genome_locations.tsv\")\n",
    "    cam2genom = {}\n",
    "    for line in open(fp, \"r\"):\n",
    "        genomeid, fp = line.strip().split(\"\\t\")\n",
    "        ncbi_id = fp.rsplit(\"/\",1)[1].rsplit(\".\",1)[0]\n",
    "        cam2genom[genomeid] = ncbi_id\n",
    "        cam2genom[ncbi_id] = genomeid\n",
    "    #for fp_dist in os.listdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "1ea07f58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Genome1': 'NC_014328_1',\n",
       " 'NC_014328_1': 'Genome1',\n",
       " 'Genome2': 'NZ_CP020566_1',\n",
       " 'NZ_CP020566_1': 'Genome2',\n",
       " 'Genome3': 'NZ_CP053893_1',\n",
       " 'NZ_CP053893_1': 'Genome3',\n",
       " 'Genome4': 'NZ_LT906445_1',\n",
       " 'NZ_LT906445_1': 'Genome4',\n",
       " 'Genome5': 'NZ_LT906470_1',\n",
       " 'NZ_LT906470_1': 'Genome5'}"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cam2genom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5e65e0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genome_sizes(combined_fa: str=\"../data/simulated_data/input_genomes/combined.fa\"):\n",
    "    genome_size = {}\n",
    "    for record in SeqIO.parse(combined_fa, \"fasta\"):\n",
    "        genome_size[record.name.replace(\".\",\"_\")] = len(record.seq)\n",
    "    return genome_size\n",
    "get_genome_sizes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "3f174bd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "datalabel = \"../data/simulated_data/camisim/0_4GB/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "15e3e51d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp = os.path.join(datalabel, \"internal/genome_locations.tsv\")\n",
    "cam2genom = {}\n",
    "for line in open(fp, \"r\"):\n",
    "    genomeid, fp = line.strip().split(\"\\t\")\n",
    "    ncbi_id = os.path.basename(fp).rsplit(\".\",1)[0]\n",
    "    cam2genom[genomeid] = ncbi_id\n",
    "    cam2genom[ncbi_id] = genomeid\n",
    "\n",
    "g_sizes = get_genome_sizes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "89736074",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>distribution</th>\n",
       "      <th>samle</th>\n",
       "      <th>ncbi</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>weighted_dist</th>\n",
       "      <th>readshare</th>\n",
       "      <th>expected_average_coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NC_014328_1</td>\n",
       "      <td>4630065</td>\n",
       "      <td>926013.0</td>\n",
       "      <td>8.238640e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_CP020566_1</td>\n",
       "      <td>2071952</td>\n",
       "      <td>414390.4</td>\n",
       "      <td>3.686788e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome3</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_CP053893_1</td>\n",
       "      <td>5984367</td>\n",
       "      <td>1196873.4</td>\n",
       "      <td>1.064846e+08</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_LT906445_1</td>\n",
       "      <td>2132142</td>\n",
       "      <td>426428.4</td>\n",
       "      <td>3.793888e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_LT906470_1</td>\n",
       "      <td>2041290</td>\n",
       "      <td>408258.0</td>\n",
       "      <td>3.632228e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NC_014328_1</td>\n",
       "      <td>4630065</td>\n",
       "      <td>926013.0</td>\n",
       "      <td>8.238640e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_CP020566_1</td>\n",
       "      <td>2071952</td>\n",
       "      <td>414390.4</td>\n",
       "      <td>3.686788e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome3</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_CP053893_1</td>\n",
       "      <td>5984367</td>\n",
       "      <td>1196873.4</td>\n",
       "      <td>1.064846e+08</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_LT906445_1</td>\n",
       "      <td>2132142</td>\n",
       "      <td>426428.4</td>\n",
       "      <td>3.793888e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_LT906470_1</td>\n",
       "      <td>2041290</td>\n",
       "      <td>408258.0</td>\n",
       "      <td>3.632228e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NC_014328_1</td>\n",
       "      <td>4630065</td>\n",
       "      <td>926013.0</td>\n",
       "      <td>8.238640e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_CP020566_1</td>\n",
       "      <td>2071952</td>\n",
       "      <td>414390.4</td>\n",
       "      <td>3.686788e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome3</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_CP053893_1</td>\n",
       "      <td>5984367</td>\n",
       "      <td>1196873.4</td>\n",
       "      <td>1.064846e+08</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_LT906445_1</td>\n",
       "      <td>2132142</td>\n",
       "      <td>426428.4</td>\n",
       "      <td>3.793888e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_LT906470_1</td>\n",
       "      <td>2041290</td>\n",
       "      <td>408258.0</td>\n",
       "      <td>3.632228e+07</td>\n",
       "      <td>17.793788</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id  distribution     samle           ncbi  genome_size  weighted_dist  \\\n",
       "0  Genome1           0.2  sample_0    NC_014328_1      4630065       926013.0   \n",
       "1  Genome2           0.2  sample_0  NZ_CP020566_1      2071952       414390.4   \n",
       "2  Genome3           0.2  sample_0  NZ_CP053893_1      5984367      1196873.4   \n",
       "3  Genome4           0.2  sample_0  NZ_LT906445_1      2132142       426428.4   \n",
       "4  Genome5           0.2  sample_0  NZ_LT906470_1      2041290       408258.0   \n",
       "0  Genome1           0.2  sample_2    NC_014328_1      4630065       926013.0   \n",
       "1  Genome2           0.2  sample_2  NZ_CP020566_1      2071952       414390.4   \n",
       "2  Genome3           0.2  sample_2  NZ_CP053893_1      5984367      1196873.4   \n",
       "3  Genome4           0.2  sample_2  NZ_LT906445_1      2132142       426428.4   \n",
       "4  Genome5           0.2  sample_2  NZ_LT906470_1      2041290       408258.0   \n",
       "0  Genome1           0.2  sample_1    NC_014328_1      4630065       926013.0   \n",
       "1  Genome2           0.2  sample_1  NZ_CP020566_1      2071952       414390.4   \n",
       "2  Genome3           0.2  sample_1  NZ_CP053893_1      5984367      1196873.4   \n",
       "3  Genome4           0.2  sample_1  NZ_LT906445_1      2132142       426428.4   \n",
       "4  Genome5           0.2  sample_1  NZ_LT906470_1      2041290       408258.0   \n",
       "\n",
       "      readshare  expected_average_coverage  \n",
       "0  8.238640e+07                  17.793788  \n",
       "1  3.686788e+07                  17.793788  \n",
       "2  1.064846e+08                  17.793788  \n",
       "3  3.793888e+07                  17.793788  \n",
       "4  3.632228e+07                  17.793788  \n",
       "0  8.238640e+07                  17.793788  \n",
       "1  3.686788e+07                  17.793788  \n",
       "2  1.064846e+08                  17.793788  \n",
       "3  3.793888e+07                  17.793788  \n",
       "4  3.632228e+07                  17.793788  \n",
       "0  8.238640e+07                  17.793788  \n",
       "1  3.686788e+07                  17.793788  \n",
       "2  1.064846e+08                  17.793788  \n",
       "3  3.793888e+07                  17.793788  \n",
       "4  3.632228e+07                  17.793788  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "dfs = []\n",
    "for fp_dist in glob.glob(os.path.join(datalabel, \"distributions/*\")):\n",
    "    df = pd.read_csv(fp_dist, sep=\"\\t\", names=[\"id\", \"distribution\"])\n",
    "    df[\"samle\"] = \"sample_\" + fp_dist.rsplit(\"_\",1)[1].replace(\".txt\",\"\")\n",
    "    df[\"ncbi\"] = [cam2genom[x] for x in df.id]\n",
    "    df[\"genome_size\"] = [g_sizes[x] for x in df.ncbi]\n",
    "    df[\"weighted_dist\"] = df.distribution * df.genome_size\n",
    "    df[\"readshare\"] = (df.weighted_dist/df.weighted_dist.sum())*(0.3*10**9)\n",
    "    df[\"expected_average_coverage\"] = df.readshare / df.genome_size\n",
    "    \n",
    "    dfs.append(df)\n",
    "df_total = pd.concat(dfs)\n",
    "df_total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "f49a8f2b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "1518760d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>distribution</th>\n",
       "      <th>samle</th>\n",
       "      <th>ncbi</th>\n",
       "      <th>genome_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NC_014328_1</td>\n",
       "      <td>4630065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_CP020566_1</td>\n",
       "      <td>2071952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome3</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_CP053893_1</td>\n",
       "      <td>5984367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_LT906445_1</td>\n",
       "      <td>2132142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_0</td>\n",
       "      <td>NZ_LT906470_1</td>\n",
       "      <td>2041290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NC_014328_1</td>\n",
       "      <td>4630065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_CP020566_1</td>\n",
       "      <td>2071952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome3</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_CP053893_1</td>\n",
       "      <td>5984367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_LT906445_1</td>\n",
       "      <td>2132142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_2</td>\n",
       "      <td>NZ_LT906470_1</td>\n",
       "      <td>2041290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NC_014328_1</td>\n",
       "      <td>4630065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_CP020566_1</td>\n",
       "      <td>2071952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome3</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_CP053893_1</td>\n",
       "      <td>5984367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_LT906445_1</td>\n",
       "      <td>2132142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>sample_1</td>\n",
       "      <td>NZ_LT906470_1</td>\n",
       "      <td>2041290</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id  distribution     samle           ncbi  genome_size\n",
       "0  Genome1           0.2  sample_0    NC_014328_1      4630065\n",
       "1  Genome2           0.2  sample_0  NZ_CP020566_1      2071952\n",
       "2  Genome3           0.2  sample_0  NZ_CP053893_1      5984367\n",
       "3  Genome4           0.2  sample_0  NZ_LT906445_1      2132142\n",
       "4  Genome5           0.2  sample_0  NZ_LT906470_1      2041290\n",
       "0  Genome1           0.2  sample_2    NC_014328_1      4630065\n",
       "1  Genome2           0.2  sample_2  NZ_CP020566_1      2071952\n",
       "2  Genome3           0.2  sample_2  NZ_CP053893_1      5984367\n",
       "3  Genome4           0.2  sample_2  NZ_LT906445_1      2132142\n",
       "4  Genome5           0.2  sample_2  NZ_LT906470_1      2041290\n",
       "0  Genome1           0.2  sample_1    NC_014328_1      4630065\n",
       "1  Genome2           0.2  sample_1  NZ_CP020566_1      2071952\n",
       "2  Genome3           0.2  sample_1  NZ_CP053893_1      5984367\n",
       "3  Genome4           0.2  sample_1  NZ_LT906445_1      2132142\n",
       "4  Genome5           0.2  sample_1  NZ_LT906470_1      2041290"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "cfbbf9c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NZ_LT906445_1': 2132142,\n",
       " 'NZ_CP053893_1': 5984367,\n",
       " 'NC_014328_1': 4630065,\n",
       " 'NZ_CP020566_1': 2071952,\n",
       " 'NZ_LT906470_1': 2041290}"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "def get_genome_sizes(combined_fa: str=\"../data/simulated_data/input_genomes/combined.fa\"):\n",
    "    genome_size = {}\n",
    "    for record in SeqIO.parse(combined_fa, \"fasta\"):\n",
    "        genome_size[record.name.replace(\".\",\"_\")] = len(record.seq)\n",
    "    return genome_size\n",
    "get_genome_sizes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "bea8ceb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp = \"../data/simulated_data/camisim/0_4GB/internal/genome_locations.tsv\"\n",
    "cam2genom = {}\n",
    "for line in open(fp, \"r\"):\n",
    "    genomeid, fp = line.strip().split(\"\\t\")\n",
    "    ncbi_id = fp.rsplit(\"/\",1)[1].rsplit(\".\",1)[0]\n",
    "    cam2genom[genomeid] = ncbi_id\n",
    "    cam2genom[ncbi_id] = genomeid   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "c7d4dd9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cam</th>\n",
       "      <th>fp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome1</td>\n",
       "      <td>/home/projects/dtu_00009/people/henspi/git/Scr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome2</td>\n",
       "      <td>/home/projects/dtu_00009/people/henspi/git/Scr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome3</td>\n",
       "      <td>/home/projects/dtu_00009/people/henspi/git/Scr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome4</td>\n",
       "      <td>/home/projects/dtu_00009/people/henspi/git/Scr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome5</td>\n",
       "      <td>/home/projects/dtu_00009/people/henspi/git/Scr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       cam                                                 fp\n",
       "0  Genome1  /home/projects/dtu_00009/people/henspi/git/Scr...\n",
       "1  Genome2  /home/projects/dtu_00009/people/henspi/git/Scr...\n",
       "2  Genome3  /home/projects/dtu_00009/people/henspi/git/Scr...\n",
       "3  Genome4  /home/projects/dtu_00009/people/henspi/git/Scr...\n",
       "4  Genome5  /home/projects/dtu_00009/people/henspi/git/Scr..."
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cam2genom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6999f5f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "g_size = get_genome_sizes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7d4acbdb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NZ_LT906445.1': 2132142,\n",
       " 'NZ_CP053893.1': 5984367,\n",
       " 'NC_014328.1': 4630065,\n",
       " 'NZ_CP020566.1': 2071952,\n",
       " 'NZ_LT906470.1': 2041290}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "c131f4ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "probs = [0.2 for i in range(5)]\n",
    "size = g_size.values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "00eb44b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "weighted_sum = sum([prob*s for prob,s in zip(probs, size)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "a58bb74e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3371963.2"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "weighted_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "83e383e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[17.793788496861414,\n",
       " 17.793788496861414,\n",
       " 17.793788496861414,\n",
       " 17.793788496861414,\n",
       " 17.793788496861414]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov = [((p)/weighted_sum)*total for s, p in zip(size, probs)]\n",
    "cov"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e597da7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "703.5178707609531"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "382f29ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9885438053811896"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total / sum(x*18 for x in g_size.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e23c239f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "300000000.0"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total = 0.3*10**9\n",
    "total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "10af1d66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17.793788496861414"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total/16859816"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
